### PR TITLE
Custom non-2xx response handling - DaprErrorResponseParser - Issue 783

### PR DESCRIPTION
**Pull request by maciek047**
# Description

This PR aims to address the problems described in issue #783. It introduces the DaprErrorResponseParser interface, which allows for setting a custom parser for the DaprHttpClient. This enables the deserialization of responses with structures different from DaprError. The changes are backward compatible, as the current logic is wrapped in the DefaultDaprErrorResponseParser.

## Issue reference
https://github.com/dapr/java-sdk/issues/783

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

File: sdk/src/main/java/io/dapr/client/DefaultDaprErrorResponseParser.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import java.nio.charset.StandardCharsets;
+
+import static io.dapr.client.ObjectSerializer.OBJECT_MAPPER;
+
+public class DefaultDaprErrorResponseParser implements DaprErrorResponseParser {
+
+    DaprError error;
+
+    @Override
+    public DaprException parse(@NotNull okhttp3.Response response) throws IOException {
```
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/DefaultDaprErrorResponseParser.java:17]: Two points here. First, can we make this interface not depend on Okhttp3? This way we can migrate to any http client in the near future.

Second point is the the http implementation will be deprecated and removed soon and gRPC will be the only implementation. Users of the SDK should not care about the protocol, except for service invocation.

So, I see two ways to proceed with this PR: make this interface protocol agnostic, making it work on with gRPC and http; Or create a new method for HTTP service invocation only and make this be used only in HTTP service invocation.

Happy to get other idea on this that would address my first two concerns.

**Review by artursouza — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/io/dapr/client/DefaultDaprErrorResponseParser.java (hunk 1 above)
Comment (maciek047) [reply sdk/src/main/java/io/dapr/client/DefaultDaprErrorResponseParser.java]: Thanks for your input, I've added changes to address the concerns raised, the interface works now both with HTTP and gRPC and is passed via the `DaprClientBuilder`.  Let me know if this is a good direction.

File: sdk/src/main/java/io/dapr/client/DaprClientGrpc.java [hunk 2]
```diff
@@ -316,6 +324,45 @@ public <T> Mono<T> invokeMethod(InvokeMethodRequest invokeMethodRequest, TypeRef
...
+          sink.success(value);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+          if (t instanceof StatusRuntimeException) {
+            StatusRuntimeException statusException = (StatusRuntimeException) t;
+            int statusCode = statusException.getStatus().getCode().value();
+            byte[] errorDetails = statusException.getStatus().getDescription() != null
+                    ? statusException.getStatus().getDescription().getBytes(StandardCharsets.UTF_8)
```
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/DaprClientGrpc.java:343]: Get encoding from our properties. See the Properties.java file.

File: sdk/src/main/java/io/dapr/client/DaprErrorResponseParser.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package io.dapr.client;
+
+import io.dapr.exceptions.DaprException;
+import java.io.IOException;
+
+public interface DaprErrorResponseParser {
+    DaprException parse(int statusCode, byte[] response) throws IOException;
```
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/DaprErrorResponseParser.java:7]: What is statusCode? These are different for gRPC or HTTP. Maybe add a new param for the protocol, so the implementation can handle both. Another option would be to have one method for gRPC and one for HTTP.

File: sdk/src/main/java/io/dapr/client/DefaultDaprErrorResponseParser.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+import static io.dapr.client.ObjectSerializer.OBJECT_MAPPER;
+
+public class DefaultDaprErrorResponseParser implements DaprErrorResponseParser {
+
+    DaprError error;
+
+    @Override
+    public DaprException parse(int statusCode, byte[] response) {
+        String errorMessage = (response == null || new String(response).isEmpty()) ? "HTTP status code: " + statusCode : new String(response, StandardCharsets.UTF_8);
```
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/DefaultDaprErrorResponseParser.java:16]: Please, reference the encoding from the Properties.java file.

File: sdk/src/main/java/io/dapr/client/DefaultDaprGrpcErrorResponseParser.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import io.dapr.exceptions.DaprException;
+import io.grpc.Status;
+
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.ExecutionException;
+
+public class DefaultDaprGrpcErrorResponseParser implements DaprErrorResponseParser {
+    @Override
+    public DaprException parse(int statusCode, byte[] errorDetails) {
+        String errorMessage = new String(errorDetails, StandardCharsets.UTF_8);
```
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/DefaultDaprGrpcErrorResponseParser.java:13]: same.

File: sdk/src/main/java/io/dapr/client/DefaultDaprErrorResponseParser.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        String errorCode = "UNKNOWN";
+        DaprException unknownException = new DaprException(errorCode, errorMessage);
+
+        if ((response != null) && (response.length != 0)) {
+            try {
+                error = OBJECT_MAPPER.readValue(response, DaprError.class);
+            } catch (IOException e) {
+                return new DaprException("UNKNOWN", new String(response, StandardCharsets.UTF_8));
```
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/DefaultDaprErrorResponseParser.java:24]: same.

File: sdk/src/test/java/io/dapr/client/DaprHttpTest.java [hunk 7]
```diff
@@ -139,7 +144,7 @@ public void invokeGetMethod() throws IOException {
     mockInterceptor.addRule()
       .get("http://127.0.0.1:3500/v1.0/get")
       .respond(serializer.serialize(EXPECTED_RESULT));
-    DaprHttp daprHttp = new DaprHttp(Properties.SIDECAR_IP.get(), 3500, okHttpClient);
+    DaprHttp daprHttp = new DaprHttp(Properties.SIDECAR_IP.get(), 3500, okHttpClient, DEFAULT_PARSER);
```
Comment (artursouza) [inline sdk/src/test/java/io/dapr/client/DaprHttpTest.java:147]: Use an overloaded version of the constructor instead. So, there will not be changes needed everywhere.

**Review by artursouza — CHANGES_REQUESTED** [review]
I like where this is going. Thanks a lot.

File: sdk/src/main/java/io/dapr/client/DaprErrorResponseParser.java (hunk 3 above)
Comment (maciek047) [reply sdk/src/main/java/io/dapr/client/DaprErrorResponseParser.java:7]: My intention with the DaprErrorResponseParser interface was to provide separate implementations for handling different protocols (as demonstrated by DefaultDaprHttpErrorResponseParser and DefaultDaprGrpcErrorResponseParser). I believe this approach maintains a clean separation of concerns between gRPC and HTTP error handling, which is why I opted not to include a protocol parameter in the interface. Ofc if you think there is a strong reason for having a single implementation handling both protocols, then I can change it

**Comment by ParrySMS** [comment]
@maciek047  Hi, is there any updated?

**Comment by ParrySMS** [comment]
Hi @maciek047 , Could you please help to add more unit test to fix the coverage issue? It will not take a long time with GPT.

**Comment by cicoyle** [comment]
gentle ping @maciek047 - mind updating this PR?

**Comment by artursouza** [comment]
Issue #783 is now resolved by https://github.com/dapr/java-sdk/pull/1009 - it is a different approach but makes the raw error payload available.

**Pull request closed without merging**